### PR TITLE
Try move-initializing in _ContiguousArrayBuffer's += when rhs is uniquely referenced

### PR DESCRIPTION
**Pull request by Catfish-Man**

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/ArrayBuffer.swift [hunk 1]
```diff
@@ -101,7 +101,7 @@ extension _ArrayBuffer {
 
   /// Returns `true` iff this buffer's storage is uniquely-referenced.
   @inlinable
-  internal mutating func isUniquelyReferenced() -> Bool {
+  internal func isUniquelyReferenced() -> Bool {
```
Comment (eeckstein) [inline stdlib/public/core/ArrayBuffer.swift]: Isn't this change ABI breaking?

Comment (atrick) [reply stdlib/public/core/ArrayBuffer.swift]: How is CoW supposed to work if we're checking uniqueness on a copy of 'self'? This all needs to work without any optimization kicking in.

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/ArraySlice.swift [hunk 2]
```diff
@@ -1265,22 +1265,15 @@ extension ArraySlice {
...
-      for x in self {
-        p.initialize(to: x)
-        p += 1
-      }
-    }
+    //TODO: _buffer._copyContents?
+    _buffer._copyContents(
+      subRange: _buffer.indices,
+      initializing: p
+    )
```
Comment (atrick) [inline stdlib/public/core/ArraySlice.swift]: Assuming this works... seems like we need to clarify in the code that, even though we're still using `self` after `_copyContents`, it's storage may point to uninitialized memory.

**Review by atrick — COMMENTED** [review]
This basically makes sense, I'm just not sure it's going to hold up in all cases with inlining.

File: stdlib/public/core/ArraySlice.swift (hunk 2 above)
Comment (Catfish-Man) [reply stdlib/public/core/ArraySlice.swift]: ah, yeah. Maybe set the count of self to 0?

**Comment by Catfish-Man** [comment]
Well that made things considerably worse :(

**Comment by Catfish-Man** [comment]
The plot thickens: it didn't!

Instead, something in the last 60 commits or so regressed this :( bisecting has been inconclusive

**Comment by atrick** [comment]
Ok, so the reason that `_isUnique` needs to be `inout` is so the optimizer does not remove any necessary retains of the reference before the check. The optimizer is under no obligation to keep the retain count of `storage` at > 1 even if a real semantic copy was originally needed.

By making a separate (unowned) copy of the reference here before checking uniqueness, you're bypassing that safeguard. So, the same optimization you're relying on to remove the copy on the caller-side coud, after inlining, remove that copy even when it's really needed (i.e. the caller continues to use `self`).

In fact, to be safe, we should have a compiler diagnostic that prevents anyone from checking uniqueness of an unowned reference...

It looks like the basic problem that you need to solve is that you need a uniqueness check on a value that isn't being mutated, so we can't refer to an address without making a new copy:

`var temp = self.storage // oops, this bumps the retain count`

I can think of a couple of ways around this:

1. Introduce a `%movedRef = Builtin.consume(%consumedRef)`. Now you can continue to do the uniqueness using an unowned reference after getting it back from the `consume`. This separates the caller's reference from the one being consumed and checked for uniqueness even after inlining.

2. Use the new `(%isUniq, %uniqRef) = Builtin.beginCowMutation(%consumedRef)`  that @eeckstein  want's to implement anyway. Then you don't need to use the `Unmanaged` API at all.

All that said, the benchmark results you're getting with this patch should be a valid upper bound on the performance you would get with a new builtin.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**